src/oc_erchef/apps/oc_chef_authz/src/oc_chef_authz_acl_constraints.erl

```diff
@@ -0,0 +1,94 @@
+%% -*- erlang-indent-level: 4;indent-tabs-mode: nil; fill-column: 92 -*-
+%% ex: ts=4 sw=4 et
+%% @author Mark  Mzyk <mm@chef.io>
+%% Copyright 2015 Chef Software, Inc.
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+
+-module(oc_chef_authz_acl_constraints).
+
+-ifdef(TEST).
+-compile([export_all]).
+-endif.
+
+-export([check_acl_constraints/4]).
+
+check_acl_constraints(AuthzId, Type, AclPerm, Ace) ->
+  check_acl_constraints(AuthzId, Type, AclPerm, Ace, acl_checks()).
+
+check_acl_constraints(AuthzId, Type, AclPerm, Ace, AclChecks) ->
+  case lists:filtermap(fun(Check) -> Check(AuthzId, Type, AclPerm, Ace) end, AclChecks) of
+    [] ->
+      ok;
+    Failures ->
+      Failures
+  end.
+
+acl_checks() ->
+  [
+    fun check_admins_group_removal_from_grant_ace/4
+  ].
+
+check_admins_group_removal_from_grant_ace(AuthzId, Type, AclPerm, NewAce) ->
+  %% It is necessary to pull the current ace and compare to the new ace.
+  %% This is because there are some groups that don't have the admin
+  %% group by default, such as billing-admins. This will have the effect
+  %% that if a group doesn't have the admin group, but then it is later added,
+  %% the admin group will never be able to be removed.
+  case AclPerm of
+    <<"grant">> ->
+      NewGroups = extract_acl_groups(AclPerm, NewAce),
+      CurrentAce = oc_chef_authz_acl:fetch(Type, AuthzId),
```

Comment:
So I looked into if we could pass in the current state, but I'm not seeing anywhere that we'd already have it before this point. Since this endpoint only allows a put, we have the new ACL state, but we don't appears to have the existing ACL state. 

The acl_state record that might contain this data is updated in the oc_chef_wm_acl_permission code to hold the new ACL, so that doesn't have what we're looking for.